### PR TITLE
Add E2E test for tracking opt-in toggle

### PR DESCRIPTION
**Pull request by circlecube**
## Summary

Addresses issue #274 

## Relevant technical choices

Adds tests to verify default opt-in value and tracking is in place as well as verify toggling opt-in setting removes tracking. 

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Comment by aaemnnosttv** [comment]
@googlebot I consent.

File: bin/local-env/install-wordpress.sh [hunk 1]
```diff
@@ -94,6 +94,10 @@ docker-compose $DOCKER_COMPOSE_FILE_OPTIONS run --rm $CONTAINER touch /var/www/h
 echo -e $(status_message "Activating Google Site Kit plugin...")
 docker-compose $DOCKER_COMPOSE_FILE_OPTIONS run --rm -u 33 $CLI plugin activate google-site-kit --quiet
 
+# Set pretty permalinks.
+echo -e $(status_message "Setting permalink structure...")
+docker-compose $DOCKER_COMPOSE_FILE_OPTIONS run --rm -u 33 $CLI rewrite structure '%postname%' --hard
```
Comment (swissspidy) [inline bin/local-env/install-wordpress.sh]: Curious: why was this needed?

If the tests fail without pretty permalinks that sounds like a bug to me.

File: tests/e2e/specs/admin-tracking.test.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	beforeAll( async() => {
+		await activatePlugin( 'google-site-kit' );
+		await activatePlugin( 'e2e-tests-auth-plugin' );
+		await resetSiteKit();
+	} );
+
+	beforeEach( async() => {
+		await visitAdminPage( 'admin.php', 'page=googlesitekit-settings' );
+		await page.waitForSelector( '.mdc-tab-scroller__scroll-content' );
+		page.click( '.mdc-tab-scroller__scroll-content button:last-child' );
```
Comment (swissspidy) [inline tests/e2e/specs/admin-tracking.test.js]: Without going through this manually, one has no idea what this clicks on.

Let's use the same approach as in #343 for this for consistency reasons:

```js
await visitAdminPage( 'admin.php', 'page=googlesitekit-settings' );
await page.waitForSelector( '.mdc-tab-scroller__scroll-content' );

// Click on Admin Settings Tab.
await expect( page ).toClick( 'button.mdc-tab', { text: 'Admin Settings' } );
```

File: tests/e2e/specs/admin-tracking.test.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+ * WordPress dependencies
+ */
+import { activatePlugin, deactivatePlugin, visitAdminPage } from '@wordpress/e2e-test-utils';
+import { URL } from 'url';
+import { resetSiteKit } from '../utils';
+
+describe( 'Providing client configuration', () => {
+
+	beforeAll( async() => {
+		await activatePlugin( 'google-site-kit' );
```
Comment (swissspidy) [inline tests/e2e/specs/admin-tracking.test.js]: The Site Kit plugin should actually be already active during E2E tests (done in the Docker setup). The only exception where it needs to be (de)activated is in the test for the plugin activation itself.

File: tests/e2e/specs/admin-tracking.test.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+	beforeEach( async() => {
+		await visitAdminPage( 'admin.php', 'page=googlesitekit-settings' );
+		await page.waitForSelector( '.mdc-tab-scroller__scroll-content' );
+		page.click( '.mdc-tab-scroller__scroll-content button:last-child' );
+	} );
+
+	afterAll( async() => {
+		await resetSiteKit();
+		await deactivatePlugin( 'google-site-kit' );
```
Comment (swissspidy) [inline tests/e2e/specs/admin-tracking.test.js]: As mentioned above, no need to deactivate the plugin.

File: tests/e2e/specs/admin-tracking.test.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+		await deactivatePlugin( 'e2e-tests-auth-plugin' );
+	} );
+
+	it( 'Should select opt-in by default', async() => {
```
Comment (swissspidy) [inline tests/e2e/specs/admin-tracking.test.js]: We currently begin the test names with lowercase letters, let's do this here as well for consistency reasons.

File: tests/e2e/specs/admin-tracking.test.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+		await page.waitForSelector( '#opt-in' );
+
+		expect( await page.$eval( '#opt-in', el => el.checked ) ).toBe( true );
```
Comment (swissspidy) [inline tests/e2e/specs/admin-tracking.test.js]: We currently successfully use `toMatchElement` to make these assertions simpler to read and write. Example here:

```js
await expect( page ).toMatchElement( '#opt-in[checked]' );
```

File: tests/e2e/specs/admin-tracking.test.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+	} );
+	it( 'Should have tracking code when opted in', async() => {
```
Comment (swissspidy) [inline tests/e2e/specs/admin-tracking.test.js]: Here there is no empty line between the tests

File: tests/e2e/specs/admin-tracking.test.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+	} );
+
+	it( 'Should uncheck opt-in box when clicked', async() => {
+
+		await page.waitForSelector( '#opt-in' );
+
+		await page.$eval( '#opt-in', elem => elem.click() );
+
+		await page.waitForResponse( res => {
```
Comment (swissspidy) [inline tests/e2e/specs/admin-tracking.test.js]: Nice choice here!

File: bin/local-env/install-wordpress.sh (hunk 1 above)
Comment (aaemnnosttv) [reply bin/local-env/install-wordpress.sh]: It's for the predicate function used in `page.waitForResponse()`, otherwise the REST url uses all query parameters which would require more parsing.

I agree that it should work either way, but we need to identify that request specifically and it both seemed easier and likely more common in the real world to use pretty permalinks.

File: tests/e2e/specs/admin-tracking.test.js (hunk 6 above)
Comment (aaemnnosttv) [reply tests/e2e/specs/admin-tracking.test.js]: That matches the html of the element when it was rendered on page load, not the current state.

If the toggle was enabled when the page loaded, it will have a `checked` attribute, but otherwise it will not. This attribute does not change when the input is toggled, that's why we're checking the property on the element.

Comment (swissspidy) [reply tests/e2e/specs/admin-tracking.test.js]: Ah, yes of course. Thanks for clarification :-)

File: tests/e2e/specs/admin-tracking.test.js (hunk 2 above)
Comment (circlecube) [reply tests/e2e/specs/admin-tracking.test.js]: Thanks for the consistency! This is updated.

File: tests/e2e/specs/admin-tracking.test.js (hunk 3 above)
Comment (circlecube) [reply tests/e2e/specs/admin-tracking.test.js]: That makes perfect sense, no longer activating/deactivating the plugin.

File: tests/e2e/specs/admin-tracking.test.js (hunk 4 above)
Comment (circlecube) [reply tests/e2e/specs/admin-tracking.test.js]: Fixed here too

File: tests/e2e/specs/admin-tracking.test.js (hunk 5 above)
Comment (circlecube) [reply tests/e2e/specs/admin-tracking.test.js]: All test names update to lower case.

File: tests/e2e/specs/admin-tracking.test.js (hunk 6 above)
Comment (circlecube) [reply tests/e2e/specs/admin-tracking.test.js]: Updated a couple that could be simplified when the checkbox did not toggle.

**Comment by circlecube** [comment]
Thanks for the thoughtful review @swissspidy - I've addressed all feedback I believe and we're just waiting for your response regarding the permalinks above.

File: bin/local-env/install-wordpress.sh (hunk 1 above)
Comment (swissspidy) [reply bin/local-env/install-wordpress.sh]: "More parsing" meaning using `reqURL.params.get()` instead of `reqURL.pathname`? 🤔 

But yeah I guess it's fine to keep this. It shouldn't bite us in the future.
